File: libschema/src/symbol/method_map.rs
use std::collections::HashMap;
use std::ops::ShlAssign;

use serde::{Deserialize, Serialize};

use crate::metadata::Metadata;
use crate::symbol::{Declaration, Function, Metable, Property};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MethodMap {
    #[serde(flatten)]
    pub declaration: Declaration,

    /// Parent inheritance if any
    pub parent: Option<String>,

    /// Functions within this methodmap
    pub methods: HashMap<String, Function>,

    /// Properties within this methodmap
    pub properties: HashMap<String, Property>,
}

impl Metable for MethodMap {
    fn metadata(&mut self) -> &mut Option<Metadata> {
        &mut self.declaration.documentation.metadata
    }
}

impl ShlAssign for MethodMap {
    fn shl_assign(&mut self, rhs: Self) {
        self.declaration <<= rhs.declaration;
        self.parent = rhs.parent;
        self.methods = rhs.methods;
        self.properties = rhs.properties;
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DPMethodMap {
    #[serde(flatten)]
    pub declaration: Declaration,

    /// Parent inheritance if any
    pub parent: Option<String>,

    /// Functions within this methodmap
    pub methods: Vec<Function>,

    /// Properties within this methodmap
    pub properties: Vec<Property>,
}

File: schema/src/interfaces/symbol/method_map.ts
import { IDeclaration } from './base';
import { IProperty } from './property';
import { IFunction } from './function';

export interface IMethodMap extends IDeclaration {
    /**
     * @brief Parent inheritance if any
     * @readonly
     */
    readonly parent?: string;

    /**
     * @brief Functions within this methodmap
     * @readonly
     */
    readonly methods: Record<string, IFunction>;

    /**
     * @brief Properties within this methodmap
     * @readonly
     */
    readonly properties: Record<string, IProperty>;
}
